Initial next gen scope support for shellscript

fidgetingbits opened the pull request:
## What

Adds support for some next gen scopes to `shellscript` programming language.

This is something I did recently while working on some bash scripts, but haven't had time to finish it off yet.

## Checklist

- [x] Recorded tests for the new language
- [x] Used `"change"` / `"clear"` instead of` "take"` for selection tests to make recorded tests easier to read
- [x] Added a few specific tests that use `"chuck"` instead of `"change"` to test removal behaviour when it's interesting, especially:
  - [x] `"chuck arg"` with single argument in list
  - [x] `"chuck arg"` with multiple arguments in list
  - [x] `"chuck item"` with single argument in list
  - [x] `"chuck item"` with multiple arguments in list
- [x] Added `@textFragment` captures. Usually you want to put these on comment and string nodes. This enables `"take round"` to work within comments and strings.
- [ ] Added a test for `"change round"` inside a string, eg `"hello (there)"`
- [-] Supported` "type"` both for type annotations (eg `foo: string`) and declarations (eg `interface Foo {}`) (and added tests for this behaviour 😊)
- [ ] Supported` "item"` both for map pairs and list entries (with tests of course)

## Scope Support

Legend: &check; Supported, &cross; Not supported, _ Not applicable

| Supported | Tested  | Term             | Capture                       | Definition                                                          | Comment |
| --------- | ------- | ---------------- | ----------------------------- | ------------------------------------------------------------------- | ------- |
| &check;   | &check; | `list`           | `@list`                       | List type equivalent                                                | -       |
| &check;   | &check; | `inside list`    | `@list.interior`              | Inside of a list                                                    | -       |
|  _         |   _      | `map`            | `@map`                        | Dictionary type equivalent                                          | -       |
|   _        |   _      | `inside map`     | `@map.interior`               | Inside of a dictionary                                              | -       |
|   _        |   _      | `key`            | `@collectionKey`              | Dictionary key equivalent                                           | -       |
| &check;   | &check; | `funk`           | `@namedFunction`              | A named function declaration                                        | -       |
| &check;   | &check; | `inside funk`    | `@namedFunction.interior`     | The inside of a lambda declaration                                  | -       |
| &check;   | &check; | `funk name`      | `@functionName`               | Name of declared function                                           | -       |
| _         | _       | `lambda`         | `@anonymousFunction`          | A lambda declaration                                                | -       |
| _         | _       | `inside lambda`  | `@anonymousFunction.interior` | The inside of a lambda declaration                                  | -       |
| &check;   | &check; | `name`           | `@name`                       | Variable name                                                       | -       |
| &check;   | &check; | `value`          | `@value`                      | Right-hand-side value of an assignment                              | -       |
|  _         | _         | `value`          | `@value`                      | Value returned from a function                                      | -       |
|           |         | `value`          | `@value`                      | Value of a key-value pair                                           | -       |
| &check;   | &check; | `state`          | `@statement`                  | Any single coded statement                                          | -       |
| &check;   | &check; | `if state`       | `@ifStatement`                | An if conditional block                                             | -       |
| &check;   | &check; | `condition`      | `@condition`                  | Condition of an if block                                            | -       |
| &check;   |        | `condition`      | `@condition`                  | Condition of a while loop                                           | -       |
| &check;          |         | `condition`      | `@condition`                  | Condition of a do while style loop                                  | -       |
| &check;   |         | `condition`      | `@condition`                  | Condition of a for loop                                             | -       |
| &check;    |         | `condition`      | `@condition`                 | Condition of a switch case                                    | -       |
|   _        |  _       | `condition`      | `@condition`                  | Condition of a ternary expression                                   | -       |
| &check;   | &check; | `branch`         | `@branch`                     | The resulting code associated with af conditional expression        | -       |
| &check;   | &check; | `inside branch`  | `@branch.interior`            | The statements associated with the body of a conditional expression | -       |
| &check;   | &check; | `comment`        | `@comment`                    | Code comment                                                        | -       |
|  _        |  _       | `comment`        | `@comment`                    | Multi-line code comment                                             | -       |
|           |         | `inside comment` | `@comment.interior`           | The inside of a code comment                                        | -       |
| &check;   | &check; | `string`         | `@string`                     | Single line strings                                                 | -       |
| _         | _       | `string`         | `@string`                     | Multi-line strings                                                  |         |
| &check;   | &check; | -                | `@textFragment`               | Used to capture string-type nodes (strings and comments)            | -       |
| &check;   | &check; | `call`           | `@functionCall`               | A function call (not a function definition)                         | -       |
| &check;   | &check; | `callee`         | `@functionCallee`             | Name of the function being called                                   | -       |
|  _         |  _       | `arg`            | `@argumentOrParameter`        | Arguments to function definition                                    | -       |
| &check;   | &check; | `arg`            | `@argumentOrParameter`        | Arguments to function call                                          | -       |
| &check;   | &check; | `arg`            | `@argumentOrParameter`        | Interpolated variables in strings                                   | -       |
| _         | _       | `class`          | `@class`                      | Class or structure declaration                                      | -       |
| _         | _       | `inside class`   | `@class.interior`             | The inside of a class declaration                                   | -       |
| _         | _       | `class name`     | `@className`                  | Name of class or structure declaration                              | -       |
| _         | _       | `type`           | `@type`                       | Type declarations                                                   | -       |
| &check;         | &check;      | `regex`           | `@regex`                       | Regular expression                                                   | -       |

pokey commented on queries/shellscript.scm:
```diff
@@ -0,0 +1,161 @@
+        .
+        (_)? @_.interior
+        .
+        "}"
```
Hmm I believe forcing these braces will break the other kinds of function definition https://github.com/tree-sitter/tree-sitter-bash/blob/7331995b19b8f8aba2d5e26deb51d2195c18bc94/grammar.js#L381-L382

pokey reviewed (commented):
Once again fantastic work. Left a few minor comments

fidgetingbits commented on queries/shellscript.scm:
```diff
@@ -0,0 +1,161 @@
+
+;;!! var="foo"
+;;!      ^^^^^
+(string) @string @textFragment @argumentOrParameter.iteration
```
There was a discussion that some of you had in another issue which I forget off hand, where iirc the consensus was that ${} expansions inside of strings were going to be referred to be as arg scopes. So I've been testing that on the shell script stuff and it seems to work pretty nicely. I can't remember offhand if I added it to nix as well, which uses similar expansions.

pokey replied:
ohh right yeah if interpolations are args, then string should be iteration scope. Feels a bit surprising to get trapped in a string for "every arg" tho. cc/ @AndreasArvidsson

AndreasArvidsson replied:
Yeah I'm a bit hesitant to make strings the iteration scope for arguments.

fidgetingbits replied:
So there are at least 3 options I guess?

1) Only add an iteration scope if the string actually has an interpolation inside, so it's less surprising 
2) Keep interpolation as args, but just don't have an iterarion scope for the container strings (I feel like being able to target every interpolation in a string could be nice though?)
3) Create a new scope and don't use args

pokey replied:
1) is an interesting possibility. We don't usually restrict depending on whether the scope actually appears, but I could see an argument here

Re 2), you don't actually need to have an iteration scope to target every interpolation in a string. You'd just say `"every arg string"`. Iteration scope is only used when there's no explicit range given for iteration. That being said, in https://github.com/cursorless-dev/cursorless/issues/2128#issuecomment-1856147061, we've started leaning towards requiring iteration scope between layers of any nested scope, so that would argue for having an iteration scope here cc/ @AndreasArvidsson 

3) is an interesting idea, and certainly I could see this being a fine-grained scope that gets rolled up into arg, tho we don't really support that yet

fidgetingbits replied:
For now I just removed the iteration scope, but keep the interpolations as args. I'll mark this unresolved in case there is some other type of iteration scope I should add in light of 2).

AndreasArvidsson commented on queries/shellscript.scm:
```diff
@@ -0,0 +1,161 @@
+;;!       ^^^^^
+(_
+    argument: (_) @argumentOrParameter
+) @_.iterator
```
Most definitely

fidgetingbits commented:
All issues from initial review pass are fixed. I have support for a few node types that are dependent on an updated tree-sitter-bash ([vscode-parse-tree PR #77](https://github.com/cursorless-dev/vscode-parse-tree/pull/77)). I've also fixed lots of other quirks and edge cases I ran into. 

I added the initial scope facet files, but running the `Update Fixtures` config throws a bunch of errors (not sure if nix-related or not), and I haven't dug into it quite yet. So I don't see any updated content in the test .scope file I created, but still pushed it anyway.

pokey commented:
I switched to tip of bash branch, and noticed some tests breaking. I'm not sure if that was the cause, but I fixed them. Along the way I cleaned up a bit and added / updated facet tests. The auto facet update worked fine for me fwiw

See my fixes; haven't looked at all the code, but possible some of the techniques there apply elsewhere? But if not limk and I'll continue reviewing

fidgetingbits commented on queries/shellscript.scm:
```diff
@@ -267,82 +267,45 @@
-)
+    value: (_) @_.removal.end.startOf
+  )
+) @_.domain @_.removal.start.startOf
```
I don't think I'd ever seen/noticed `.removal`. Much nicer :D

pokey replied:
Yeah we generally prefer leading / trailing if it makes sense, but note that the behaviour of leading / trailing may not be as you expect. During removal, we only remove one of them. The canonical example is the comma in a comma-separated list: you don't want to remove both commas

If you need to remove something from both sides, removal is the way to go

pokey commented:
> How do I actually run this for a subset? There is very little documentation for the regular expressions

See https://www.cursorless.org/docs/contributing/#running-a-subset-of-tests

Thought the regex was documented somewhere. It’s just the one supported by mocha cli. A PR to add a link to that to those docs would be much appreciated ☺️ 

(pasted from slack form posterity)

> the example folders don't seem to actually exist.

Sorry, not sure I follow. Which example folders?

fidgetingbits commented:
My problem was the regex. I realized after I ran the non subset and it worked. My regex was set to languages/shellscript because that's what I had been doing for the test cases (not sure why I prefixed languages/ in the first place), and of course scope doesn't have the languages/ subdir.

Fwiw, one error I get when running tests (though doesn't impact the scopes updating):

```
rejected promise not handled within 1 second: NeedsInitialTalonUpdateError: Custom spoken forms file not found at /tmp/ffa46ea4b829416627bdcd0114572b73/state.json. Using default spoken forms.
extensionHostProcess.js:131
stack trace: NeedsInitialTalonUpdateError: Custom spoken forms file not found at /tmp/ffa46ea4b829416627bdcd0114572b73/state.json. Using default spoken forms.
	at TalonSpokenFormsJsonReader.getSpokenFormEntries (/home/aa/persist/source/fidgetingbits-cursorless/packages/cursorless-vscode/dist/extension.cjs:50288:15)
	at async CustomSpokenForms.updateSpokenFormMaps (/home/aa/persist/source/fidgetingbits-cursorless/packages/cursorless-vscode/dist/extension.cjs:34339:26)
```

Not sure if this is another issue with TMPDIR style stuff or what. Didn't get time to dig into it yet.

pokey commented:
Is your cursorless talon up to date?

fidgetingbits commented:
I'll take a look at it tomorrow to see if I spot any candidates for using removal, and also add the fixtures I've been working on.

fidgetingbits commented:
Trying to update this in light of the conflict from getLanguageScopeSupport being deleted, and seeing new pre-commit errors I've not run into before:

```
eslint...................................................................Passed
prettier.................................................................Failed
- hook id: prettier
- exit code: 1

[error] Cannot find package 'prettier-plugin-tailwindcss' imported from /home/aa/dev/talon/fidgetingbits-curs
orless/noop.js
[error] Cannot find package 'prettier-plugin-tailwindcss' imported from /home/aa/dev/talon/fidgetingbits-curs
orless/noop.js
[error] Cannot find package 'prettier-plugin-tailwindcss' imported from /home/aa/dev/talon/fidgetingbits-curs
orless/noop.js
[error] Cannot find package 'prettier-plugin-tailwindcss' imported from /home/aa/dev/talon/fidgetingbits-curs
orless/noop.js
[error] Cannot find package 'prettier-plugin-tailwindcss' imported from /home/aa/dev/talon/fidgetingbits-curs
orless/noop.js
[error] Cannot find package 'prettier-plugin-tailwindcss' imported from /home/aa/dev/talon/fidgetingbits-curs
orless/noop.js
[error] Cannot find package 'prettier-plugin-tailwindcss' imported from /home/aa/dev/talon/fidgetingbits-curs
orless/noop.js
[error] Cannot find package 'prettier-plugin-tailwindcss' imported from /home/aa/dev/talon/fidgetingbits-curs
orless/noop.js
[error] Cannot find package 'prettier-plugin-tailwindcss' imported from /home/aa/dev/talon/fidgetingbits-curs
orless/noop.js
[error] Cannot find package 'prettier-plugin-tailwindcss' imported from /home/aa/dev/talon/fidgetingbits-curs
orless/noop.js
[error] Cannot find package 'prettier-plugin-tailwindcss' imported from /home/aa/dev/talon/fidgetingbits-curs
orless/noop.js

format-recorded-tests....................................................Failed
- hook id: format-recorded-tests
- exit code: 1

✘ [ERROR] Could not resolve "chai"

    packages/common/src/testUtil/runRecordedTest.ts:34:23:
      34 │ import { assert } from "chai";
         ╵                        ~~~~~~

  You can mark the path "chai" as external to exclude it from the bundle, which will remove this error.
```

Will look into it, but just throwing it out here in case you happen to know what to do.

fidgetingbits commented:
The tailwind bug is from commit de37ebf77834afad9a102cd5a4235886c72ece65. This adds an external pnpm package dependency that pre-commit itself won't install, so pre-commit install is no longer enough. Nixos doesn't have a prettier-plugin-tailwindcss npm package, so atm I have to manually pnpm install -w -D prettier-plugin-tailwindcss, which then changes the pnpm-lock.yaml file lol. :/ I will create a nix package for prettier-plugin-tailwindcss and get it upstreamed, but in the meantime I'll have to use a bit of a hack to get that to work I guess (will just install the package in the dev shell and then git restore the lock file..)

The second bug seems to be related to me recently changing the flake.nix file to use corepack per auscompgeek suggestion. Reverting it back to using hardcoded nodejs 18 makes it go away, so I will revert that latest commit in  I guess, as using pre-commit wasn't something I had tested.

fidgetingbits commented:
Looks like a bunch of tests are breaking related to @dummy and other things which I vaguely recall fixing a long time ago, so not sure what's up, but will investigate.
